pep-0544.txt

```diff
@@ -0,0 +1,967 @@
+of the protocol.
+
+The attributes (variables and methods) of a protocol that are mandatory
+for other class in order to be considered a structural subtype are called
+protocol members.
+
+
+.. _definition:
+
+Defining a protocol
+-------------------
+
+Protocols are defined by including a special new class ``typing.Protocol``
+(an instance of ``abc.ABCMeta``) in the base classes list, preferably
+at the end of the list. Here is a simple example::
+
+  from typing import Protocol
+
+  class SupportsClose(Protocol):
+      def close(self) -> None:
+          ...
+
+Now if one defines a class ``Resource`` with a ``close()`` method that has
+a compatible signature, it would implicitly be a subtype of
+``SupportsClose``, since the structural subtyping is used for
+protocol types::
+
+  class Resource:
+      ...
+      def close(self) -> None:
+          self.file.close()
+          self.lock.release()
+
+Apart from few restrictions explicitly mentioned below, protocol types can
+be used in every context where a normal types can::
+
+  def close_all(things: Iterable[SupportsClose]) -> None:
+      for t in things:
+          t.close()
+
+  f = open('foo.txt')
+  r = Resource()
+  close_all([f, r])  # OK!
+  close_all([1])     # Error: 'int' has no 'close' method
+
+Note that both the user-defined class ``Resource`` and the built-in
+``IO`` type (the return type of ``open()``) are considered subtypes of
+``SupportsClose``, because they provide a ``close()`` method with
+a compatible type signature.
+
+
+Protocol members
+----------------
+
+All methods defined in the protocol class body are protocol members, both
+normal and decorated with ``@abstractmethod``. If some or all parameters of
+protocol method are not annotated, then their types are assumed to be ``Any``
+(see PEP 484). Bodies of protocol methods are type checked, except for methods
+decorated with ``@abstractmethod`` with trivial bodies. A trivial body can
+contain a docstring. Example::
+
+  from typing import Protocol
+  from abc import abstractmethod
+
+  class Example(Protocol):
+      def first(self) -> int:     # This is a protocol member
+          return 42
+
+      @abstractmethod
+      def second(self) -> int:    # Method without a default implementation
+          """Some method."""
+
+Note that although formally the implicit return type of a method with
+a trivial body is ``None``, type checker will not warn about above example,
+such convention is similar to how methods are defined in stub files.
+Static methods, class methods, and properties are equally allowed
+in protocols.
+
+To define a protocol variable, one must use PEP 526 variable
+annotations in the class body. Additional attributes *only* defined in
+the body of a method by assignment via ``self`` are not allowed. The rationale
+for this is that the protocol class implementation is often not shared by
+subtypes, so the interface should not depend on the default implementation.
```

Comment:
I'm not adamantly against protocol variables, but I think we should at least settle what to do about read-only attributes. When you use `@property` (not using the `@f.setter` syntax) or use PEP 526 syntax inside a`NamedTuple` you get a read-only attribute; otherwise PEP 526 syntax gets you a writable attribute. Which should be the default for Protocols, and how should you be able to declare the other kind?

Note that this is *not* about immutability -- that's in the programmer's head (or perhaps guided by whether `__hash__` is defined). This is purely about whether attribute assignment should be allowed.